---
title: "Notes"
output: html_document
date: "2023-07-02"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## read.vcf()

```{r, echo=T, eval=F}
my_vcf<-read_vcf(x="inst/example.vcf", only.bi=T)
#to do:
#1. add option to only keep certain individuals with a parameter "id.range" where c(x, y), x is first individual in the range and y is last individual in the range.(e.g. first individual column in the VCF is 1, second is 2, and so on.). If the length of the vector is >2, this will be interpreted as a vector of ID positions to keep.
#2. include possibility for user to only keep certain positions with a parameter "pos.range" where c(x, y), x is beginning position and y is end position of range.
#3. instead of simply creating a data.tbale, creates a list with first element the data.table. Other elements: IDs: holds the ID names, Geno: holds the genotypes, var_ids: holds variant IDs. (low priority)
```

## parse.vcf()

```{r, echo=T, eval=F}

my_parsed_vcf<-parse_vcf(infile="inst/example.vcf", n0=108,  type="ncd1")
#CHANGES:
#1. removed most parameters
# added: assertthat::assert_that(ncd_type %in% c("ncd1", "ncd2"), msg="ncd_type must be either ncd1 or ncd2.\n")
#2. fixed many details in count_alleles and split_geno
#To do:
#2. include possibility for "infile" to be a vcf object read in with read.vcf. Might require creating a function that assesses whether an R object has the characteristics of a vcf file. 
#a copy of "my_parsed_vcf" is included as an example file in the package and is called ncd1_input

```



