---
title: "Tutorial"
output:
  xaringan::moon_reader:
    seal: false
    transition: slide
    css: xaringan-themer.css
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
      ratio: 16:9
header-includes:
  \usepackage{fvextra}
  \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}  
---
```{r setup, include=FALSE}
library(knitr)
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(
  fig.width=9, fig.height=4, fig.retina=4,
  out.width = "100%",
  cache = FALSE,
  echo = TRUE,
  message = FALSE, 
  warning = FALSE,
  hiline = TRUE
)
```

```{r xaringan-themer, include=FALSE, warning=FALSE}
library(xaringan)
library(xaringanthemer)
style_duo_accent(primary_color = "#046C9A", 
          secondary_color = "#C27D38",
          text_bold_color = "#C27D38",
          title_slide_text_color = "#C6CDF7",
          title_slide_background_color = "#D8A499",
          text_font_size=2.5,
          header_font_google = google_font("Aboreto"),
          text_font_google   = google_font("Roboto", "400", "400i"),
          code_font_google   = google_font("Fira Mono"),
          code_font_size = 0.8,
          colors = c(
  pink = "#E6A0C4",
  plum = "#9986A5",
  cornflowerblue = "#7294D4",
  weirdbrown = "#C27D38",
  lilac="#C6CDF7",
  gray="#8D8680",
  green="#78B7C5",
  orange="#F2AD00",
  ice="#D9D0D3"
)
)
extra_css <- list(
  ".small" = list("font-size" = "80%"),
  ".large" = list("font-size" = "120%"),
  ".LARGE" = list("font-size" = "150%"),
  ".full-width" = list(
    display = "flex",
    width   = "100%",
    flex    = "1 1 auto"
  )
)
```

<style>
.inverse {
  background-color: #046C9A;
  color: #D9D0D3;
}

</style>
<style type="text/css">
.remark-slide-content {
    font-size: 30px;
    padding: 1em 3em 1em 3em;
}
ul {
  list-style: none; /* Remove default bullets */
}

ul li::before {
  content: "\2022";  /* Add content: \2022 is the CSS Code/unicode for a bullet */
  color: #C27D38; /* Change the color */
  font-weight: bold; /* If you want it to be bold */
  display: inline-block; /* Needed to add space between the bullet and the text */
  width: 1em; /* Also needed for space (tweak if needed) */
  margin-left: -1em; /* Also needed for space (tweak if needed) */
}

.remark-code, .remark-inline-code { font-family: 'Source Code Pro', 'Lucida Console', Monaco, monospace;
                                    font-size: 85%;
}
</style>

---
### Install `balselr`

```{r, echo=FALSE}
library(knitr)
library(devtools)
devtools::install_github("bitarellolab/balselr")
library(balselr)
hook_output <- knit_hooks$get("output")
knit_hooks$set(output = function(x, options) {
  lines <- options$output.lines
  if (is.null(lines)) {
    return(hook_output(x, options))  # pass to default hook
  }
  x <- unlist(strsplit(x, "\n"))
  more <- "..."
  if (length(lines)==1) {        # first n lines
    if (length(x) > lines) {
      # truncate the output, but add ....
      x <- c(head(x, lines), more)
    }
  } else {
    x <- c(more, x[lines], more)
  }
  # paste these lines together
  x <- paste(c(x, ""), collapse = "\n")
  hook_output(x, options)
})
library(devtools)
library(balselr)
devtools::load_all()
``` 


```{r, echo=TRUE, eval=F}
library(devtools)
devtools::install_github("bitarellolab/balselr")
library(balselr)
```

---

### Walkthrough example

* In this example, the VCF (Variant Call Format) file contains population-level information for 108 individuals for the ingroup (pop0) and one individual for the outgroup (pop1). 
* `ncd1` only requires ingroup information while `ncd2` requires both.

---

### Install `balselr`
```{r, echo=TRUE, eval=F}
library(devtools)
devtools::install_github("bitarellolab/balselr")
library(balselr)
```

---

### Read in a `vcf` file

```{r, echo=T, eval=F}
#showing first 6 columns
read_vcf(x=system.file(package="balselr", "example.vcf"))
```

```{r, echo=F, eval=T,output.lines=8}
test<-read_vcf(x=system.file(package="balselr", "example.vcf"))[,1:6]
print(test)
```


---

### Make an `ncd1` input file


```{r, output.lines=8}
parse_vcf(
  infile = system.file(package = "balselr", "example.vcf"),
  n0 = 108, type = "ncd1")
```

---


### Make an `ncd2` input file


```{r, output.lines=8}
parse_vcf(
  infile = system.file(package = "balselr", "example.vcf"),
  n0 = 108, n1 =1, type = "ncd2")
```

---

### Run `ncd1`

```{r, output.lines=8}
ncd1(x = ncd1_input,
        tf = 0.5,
        w = 3000,
        ncores = 2,
        minIS = 8)
```

---

### Run `ncd2`

```{r, output.lines=8}
ncd2(x = ncd2_input,
        tf = 0.5,
        w = 3000,
        ncores = 2,
        minIS = 8)
```


---

### Core Functions

`read_vcf`: Read in a VCF (Variant Call Format) file. This uses `data.table` and is very fast.

`parse_vcf`: Read in a VCF (Variant Call Format) file and prepare input files for `ncd1()` and `ncd2()`.

`ncd1`: Calculate non-central statistic (NCD1). Returns a data.table object with columns:  Win.ID, S (SNPs), IS (informative sites), tf, ncd1. 

`ncd2`: Calculate non-central statistic (NCD2).Returns a data.table object with columns:  Win.ID, S (SNPs), Subst (Fixed differences), IS (informative sites), tf, ncd1. 

---

Arguments for `ncd1` and `ncd2`:

- `tf`: Target frequency.
- `w`: Window size in bp. Default 3000.
- `minIS`: Minimum number of informative sites. Default is 2. Windows with less informative sites than this threshold are discarded.
- `by`: Define how to scan the genome. "POS" (default) defined sliding windows based on `w`. "IS" defined windows around each informative site.




