#' Parse VCF files generated by SLiM v3
#'
#' @param infile The path and name of a vcf file - single chromosome
#' @param outfile The path and name for the outfile. If not provided,
#' this will be  a timestamp + infile in the current directory.
#' @param n0 Number of individuals in pop0
#' @param n1 Number of individuals in pop1
#' @param type Which input format will be generated. Options: ncd1, ncd2.
#'
#' @return Returns a data table object.
#' @export
#'
#' @examples parse_vcf(infile=system.file(package="balselr", "example.vcf"), n0=108, type="ncd1")
#' @examples parse_vcf(infile=system.file(package="balselr", "example.vcf"), n0=108, n1=2, type="ncd2")


parse_vcf <- function(infile = NULL,
                      outfile = NULL,
                      n0 = NULL,
                      n1 = NULL,
                      type = NULL) {

        assertthat::assert_that(file.exists(infile),
                                msg = glue::glue("VCF file {infile} does not exist.\n"))
        type <- tolower(type)
        assertthat::assert_that(type %in% c("ncd1", "ncd2"), msg = "ncd_type must be either ncd1 or ncd2.\n")
        if (type == "ncd2") {
                assertthat::assert_that(is.null(n1) == FALSE, msg = "n1 cannot be 'NULL'. NCD2 requires an outgroup.")
        }
        local.path = getwd()
        if (!(is.null(outfile))) {
                pat = stringr::str_split(infile,
                                         pattern = "/",
                                         simplify = T)[length(stringr::str_split(
                                                 infile,
                                                 pattern = "/",
                                                 simplify = T
                                         ))]
                outfile = outfile_path(glue::glue("{local.path}/{pat}_{type}.out"))
                #if (verbose == T) {
                #cat(glue::glue("No outfile provided. Will write this into tmp file {outfile}_{type}.out"),"\n")
                #}
                #} else{
                outfile <- glue::glue("{local.path}/{outfile}")
        }
        #}
        #}else{
        #        cat(glue::glue("No outfile provided.
        #}
        #pref <- gsub(".vcf", "", infile)
        inp <- read_vcf(x = infile)
        nind<-c(n0,n1)
        index.col <- which(colnames(inp) == "FORMAT") + 1
        # if (verbose == T) {
        #cat(glue::glue("First genotype column is {index.col}"),"\n")
        #}
        #if (verbose == T) {
        #        cat(glue::glue("Creating input file for {type} from {infile}..."),"\n")
        #}

        #
        if (type == "ncd2") {
                res <-
                        .vcf_ncd2(
                                x = inp,
                                outfile = outfile,
                                nind = nind,
                                index.col = index.col
                                #verbose = verbose
                        )
        } else if (type == "ncd1") {
                res <-
                        .vcf_ncd1(
                                x = inp,
                                outfile = outfile,
                                nind = nind,
                                index.col = index.col
                                #verbose = verbose
                        )
        }
        #      if (verbose == T) {
        #       cat(
        #               glue::glue(
        #                      "Finished making input file for {type}. File written to: {outfile}"
        #               ),
        #               "\n"
        #      )
        #
        #      } else {
        #       cat(glue::glue("File written to: {local.path}/{outfile}"), "\n")
        #    }

        #     if (intern == T) {
        ##     }
        return(res)
}
